cmd/snap-confine/snap-confine-args.c

```diff
@@ -0,0 +1,221 @@
+			args->is_classic_confinement = true;
+		} else {
+			// Report unhandled option switches
+			err = sc_error_init(SC_ARGS_DOMAIN, SC_ARGS_ERR_USAGE,
+					    "unrecognized command line option: %s",
+					    argv[optind]);
+			goto out;
+		}
+	}
+
+	// Parse positional arguments.
+	//
+	// NOTE: optind is not reset, we just continue from where we left off in
+	// the loop above.
+	for (; optind < argc; ++optind) {
+		if (args->security_tag == NULL) {
+			// The first positional argument becomes the security tag.
+			if (ignore_first_tag) {
+				// Unless we are called as ubuntu-core-launcher, then we just
+				// swallow and ignore that security tag altogether.
+				ignore_first_tag = false;
+				continue;
+			}
+			args->security_tag = strdup(argv[optind]);
+			if (args->security_tag == NULL) {
+				die("cannot allocate memory for security tag");
+			}
+		} else if (args->executable == NULL) {
+			// The second positional argument becomes the executable name.
+			args->executable = strdup(argv[optind]);
+			if (args->executable == NULL) {
+				die("cannot allocate memory for executable name");
+			}
+			// No more positional arguments are required.
+			// Stop the parsing process.
+			break;
+		}
+	}
+
+	// Verify that all mandatory positional arguments are present.
+	// Ensure that we have the security tag
+	if (args->security_tag == NULL) {
+		err = sc_error_init(SC_ARGS_DOMAIN, SC_ARGS_ERR_USAGE,
+				    "application or hook security tag was not provided");
+		goto out;
+	}
+	// Ensure that we have the executable name
+	if (args->executable == NULL) {
+		err = sc_error_init(SC_ARGS_DOMAIN, SC_ARGS_ERR_USAGE,
+				    "executable name was not provided");
+		goto out;
+	}
+
+	int i;
+ done:
+	// "shift" the argument vector left, except for argv[0], to "consume" the
+	// arguments that were scanned / parsed correctly.
+	for (i = 1; optind + i < argc; ++i) {
+		argv[i] = argv[optind + i];
+	}
+	argv[i] = NULL;
+
+	// Write the updated argc back, argv is never modified.
+	*argcp = argc - optind;
+
+ out:
+	// Don't return anything in case of an error.
+	if (err != NULL) {
+		sc_cleanup_args(&args);
+	}
+	// Forward the error and return
+	sc_error_forward(errorp, err);
+	return args;
+}
+
+void sc_args_free(struct sc_args *args)
+{
+	if (args != NULL) {
+		free(args->security_tag);
+		args->security_tag = NULL;
+		free(args->executable);
+		args->executable = NULL;
+		free(args);
+	}
+}
+
+void sc_cleanup_args(struct sc_args **ptr)
+{
+	sc_args_free(*ptr);
+	*ptr = NULL;
+}
+
+bool sc_args_is_version_query(struct sc_args *args)
+{
+	if (args == NULL) {
+		die("cannot obtain version query flag from NULL argument parser");
+	}
+	return args->is_version_query;
+}
+
+bool sc_args_is_classic_confinement(struct sc_args * args)
+{
+	if (args == NULL) {
+		die("cannot obtain classic confinement flag from NULL argument parser");
+	}
+	return args->is_classic_confinement;
+}
+
+char *sc_args_security_tag(struct sc_args *args)
```

Comment:
Do you want callers of this function to be able to modify the security_tag string? I think returning a `const char *` would be cleaner.